### ISSUE TITLE
Views don't seem to work

### ISSUE THREAD

**Issue by githubnemo** [mcf06/theano_ctc#11]
`tests/test_keras.py` suggests to use `.dimshuffle((1,0,2))` to match the expected input shape of warp-ctc. I found that this leads to wrong results.

Test file: https://gist.github.com/githubnemo/5553d3c9c0735c2f372ede2a119d4495

Suppose I use the keras compatible format `(batch, seqlen, units)` and create a tensor from it:

```
nacts = [...] # from test_ctc.py extended by one batch, shape = (3,4,5)
acts = # nacts reshaped to (4,3,5)
tsActsN = theano.shared(acts)
tsActs = tsActsN.dimshuffle((1,0,2))
loss = ctc(tsActs)
```

the result will not be the same as if I would have used `nacts` to create the `tsActs` tensor directly.
However, copying works:

```
nacts = [...] # from test_ctc.py extended by one batch, shape = (3,4,5)
acts = # nacts reshaped to (4,3,5)
tsActsN = theano.shared(acts)
tsActs = theano.shared(np.zeros((3,4,5), dtype='float32'))
for i in range(4):
    tsActs = T.set_subtensor(tsActs[:,i,:], tsActsN[i,:,:])
loss = ctc(tsActs)
```

I'm not sure how this comes but `.dimshuffle()` creates a view which might upset things.

This behavior is the same regardless of `device=cpu` or `gpu`.

**Comment by githubnemo** [comment]
I'm pretty sure this is due to the fact that `dimshuffle` only creates a view. When calling `CudaNdarray_DEV_DATA(%(acts)s);` the pointer to the raw data is returned and since `dimshuffle` only modifies the dimension, not the data, it has basically no effect.

To fix this this I can think of these options:
- Filtering views (`x->data_allocated == 0`) and telling the user to allocate themselves, thus preserving the speed-up for users that have matching input dimensions
- Copying the data with explicit dimensions to format the data on the device correctly
- Do both depending on the case

**Comment by mcf06** [comment]
I can reproduce this with the following change to tests/test_ctc.py. Replace:

tsActs = theano.shared(acts, "acts")   # float32, so should be moved to GPU

with: 

acts = acts.transpose((1,0,2)).copy()
tsActs = theano.shared(acts, "acts")   # float32, so should be moved to GPU
tsActs = tsActs.dimshuffle((1,0,2))

These should be the same -- one transpose in numpy, one in theano -- but the CTC costs are wrong. 

Annoyingly, passing tsActs.copy() to ctc_cost() doesn't work, because the Elemwise{Identity} gets optimized away. Instead, I've added some C code to deep copy the tensor if it's not C-contiguous.

**Comment by githubnemo** [comment]
That patch looks much nicer than what I could hack together in that time, neat.

My test model loss converges equally (epsilon=0.01) compared to the workaround so I would say this is fixed (tests run as well).

**Comment by nouiz** [comment]
Could it be that the CTC ops overwrite some of its inputs? Or return a view
of its inputs?

If that is the case, then this must be declared in the Op. Currently, there
is no such declaration:

http://deeplearning.net/software/theano/extending/inplace.html

**Comment by mcf06** [comment]
Actually, the outputs are freshly allocated.
